cve_bin_tool/error_handler.py

```diff
@@ -0,0 +1,133 @@
+import sys
+from enum import Enum
+from logging import Logger
+
+from rich.console import Console
+from rich.traceback import Traceback
+
+CONSOLE = Console(file=sys.stderr)
+
+
+class InsufficientArgs(Exception):
+    """ Insufficient command line arguments"""
+
+
+class InvalidCsvError(Exception):
+    """ Given File is an Invalid CSV """
+
+
+class MissingFieldsError(Exception):
+    """ Missing needed fields """
+
+
+class InvalidJsonError(Exception):
+    """ Given File is an Invalid JSON """
+
+
+class EmptyCache(Exception):
+    """
+    Raised when NVD is opened when verify=False and there are no files in the
+    cache.
+    """
+
+
+class CVEDataForYearNotInCache(Exception):
+    """
+    Raised when the CVE data for a year is not present in the cache.
+    """
+
+
+class CVEDataForCurlVersionNotInCache(Exception):
+    """
+    Raised when the CVE data for a curl version is not present in the cache.
+    """
+
+
+class AttemptedToWriteOutsideCachedir(Exception):
+    """
+    Raised if we attempted to write to a file that would have been outside the
+    cachedir.
+    """
+
+
+class SHAMismatch(Exception):
+    """
+    Raised if the sha of a file in the cache was not what it should be.
+    """
+
+
+class ExtractionFailed(ValueError):
+    """ Extraction fail """
+
+
+class UnknownArchiveType(ValueError):
+    """ Unknown archive type"""
+
+
+class ErrorMode(Enum):
+    Ignore = 0
+    NoTrace = 1
+    TruncTrace = 2
+    FullTrace = 3
+
+
+def excepthook(exc_type, exc_val, exc_tb):
+    trace = Traceback.from_exception(exc_type, exc_val, exc_tb)
+    CONSOLE.print(trace)
+    if ERROR_CODES.get(exc_type):
+        sys.exit(ERROR_CODES[exc_type])
+
+
+class ErrorHandler:
+    """
+    @summary: Error handler context manager.
+    usecases:
+        Different modes that can ignore error, print full trace, truncated trace and no trace.
+        Log messages if logger specified.
+    @param: mode (ErrorMode): Can take any valid ErrorMode as an arg and change output according to that.
+    @param: logger: logs error message specified while raising Exception if logger is passed
+                    while class initialization.
+                    Ex: raise ValueError("file required") will log 'file required'.
+    """
```

Comment:
New ErrorHandler context manager. It supports 4 types of mode Ignore (ignore error), NoTrace (trace won't print), TruncTrace(truncated trace for the context only (default)), FullTrace (full trace default when debugging.)

Reply:
While using ErrorHandler context manager if error get throws it sets exit_code to error_code of that error. This will be helpful in Ignore mode to inspect if exception get throw.